File: ruby/armstrong-numbers/armstrong_numbers.rb
module ArmstrongNumbers
  module_function

  def include?(number)
    sum =
      number
      .digits
      .then(&method(:sum_powers))

    sum == number
  end

  def sum_powers(digits)
    digits
      .each
      .with_object(digits.length)
      .sum(&method(:power))
  end
  private_class_method :sum_powers

  def power((digit, length))
    digit**length
  end
  private_class_method :power
end

File: typescript/armstrong-numbers/armstrong-numbers.ts
type PowerReducer = (acc: number, digit: number) => number

export default abstract class ArmstrongNumbers {
  static isArmstrongNumber(num: number): boolean {
    const sumOfPowers: number =
      ArmstrongNumbers.sumPowers(ArmstrongNumbers.digits(num))

    return num === sumOfPowers
  }

  private static digits(num: number): number[] {
    return Array.from(num.toString()).map(Number)
  }

  private static sumPowers(digits: number[]): number {
    return digits.reduce(ArmstrongNumbers.power(digits.length), 0)
  }

  private static power(length: number): PowerReducer {
    return (acc: number, digit: number): number => {
      return acc + Math.pow(digit, length)
    }
  }
}
